File: intervew_prep/warm_up_challenges/jumping_on_the_clouds.go
package main

import "fmt"

func findPath(clouds []int32, currentCloud int32, path []int32) []int32 {
	if len(clouds) == 1 {
		return append(path, currentCloud)
	}
	var tempPath1 []int32
	var tempPath2 []int32

	if len(clouds) > 1 && clouds[1] == 0 {
		tempPath1 = findPath(clouds[1:], currentCloud+1, append(path, currentCloud))
	}
	if len(clouds) > 2 && clouds[2] == 0 {
		tempPath2 = findPath(clouds[2:], currentCloud+2, append(path, currentCloud))
	}
	if tempPath1 == nil {
		return tempPath2
	}
	if tempPath2 == nil {
		return tempPath1
	}
	if len(tempPath1) < len(tempPath2) {
		return tempPath1
	}
	return tempPath2
}

func main() {
	// 0,0,0,0,0
	l := []int32{0, 0, 1, 0, 0, 1, 0}
	fmt.Println(findPath(l, 0, []int32{}))
}

File: intervew_prep/warm_up_challenges/jumping_on_the_clouds.exs
# Jumping on the clouds
#
defmodule CloudWatch do
  def find_path([_cloud], current, path) do
    [current | path]
  end

  def find_path([], _current, path) do
    nil
  end

  def find_path(["1" | _rest], _current, _path) do
    nil
  end

  def find_path(clouds, current, path) do
    shortest_path(
      find_path(Enum.slice(clouds, 1..(length(clouds) - 1)), current + 1, [current | path]),
      find_path(Enum.slice(clouds, 2..(length(clouds) - 1)), current + 2, [current | path])
    )
  end

  def shortest_path(nil, path2) do
    path2
  end

  def shortest_path(path1, nil) do
    path1
  end

  def shortest_path(path1, path2) when length(path1) == 0 do
    path2
  end

  def shortest_path(path1, path2) when length(path2) == 0 do
    path1
  end

  def shortest_path(path1, path2) when length(path1) <= length(path2) do
    path1
  end

  def shortest_path(path1, path2) when length(path2) < length(path1) do
    path2
  end

  def final_answer(path) do
    length(path) - 1
  end
end

IO.read(:stdio, :line)

IO.read(:stdio, :line)
|> String.split()
|> CloudWatch.find_path(0, [])
|> CloudWatch.final_answer()
|> IO.puts()
